File: Python/lld-examples/medium/restaurant-management-system/order.py
from typing import List
from enum import Enum
from datetime import datetime
import uuid
from menu import MenuItem

class OrderStatus(Enum):
    """Order status enum"""
    PENDING = "Pending"
    PREPARING = "Preparing"
    READY = "Ready"
    SERVED = "Served"
    PAID = "Paid"
    CANCELLED = "Cancelled"

class OrderItem:
    """Order item class"""
    
    def __init__(self, menu_item: MenuItem, quantity: int):
        self.menu_item = menu_item
        self.quantity = quantity
        self.unit_price = menu_item.price
        self.special_instructions = ""
    
    def get_total(self) -> float:
        """Get total price for this item"""
        return self.unit_price * self.quantity
    
    def __str__(self):
        return f"{self.menu_item.name} x{self.quantity} - ${self.get_total():.2f}"

class Order:
    """Order class representing a customer order"""
    
    def __init__(self, table, customer):
        self.order_id = str(uuid.uuid4())[:8].upper()
        self.table = table
        self.customer = customer
        self.items: List[OrderItem] = []
        self.status = OrderStatus.PENDING
        self.order_time = datetime.now()
        self.special_requests = ""
    
    def add_item(self, menu_item: MenuItem, quantity: int, special_instructions: str = ""):
        """Add item to order"""
        order_item = OrderItem(menu_item, quantity)
        if special_instructions:
            order_item.special_instructions = special_instructions
        self.items.append(order_item)
    
    def remove_item(self, menu_item: MenuItem):
        """Remove item from order"""
        self.items = [item for item in self.items if item.menu_item.item_id != menu_item.item_id]
    
    def get_total(self) -> float:
        """Get total order amount"""
        return sum(item.get_total() for item in self.items)
    
    def get_total_with_tax(self, tax_rate: float = 0.08) -> float:
        """Get total with tax"""
        subtotal = self.get_total()
        return subtotal * (1 + tax_rate)
    
    def update_status(self, status: OrderStatus):
        """Update order status"""
        self.status = status
    
    def mark_as_paid(self):
        """Mark order as paid"""
        self.status = OrderStatus.PAID
    
    def cancel_order(self):
        """Cancel order"""
        self.status = OrderStatus.CANCELLED
    
    def get_formatted_order_time(self) -> str:
        """Get formatted order time"""
        return self.order_time.strftime("%Y-%m-%d %H:%M:%S")
    
    def __str__(self):
        return f"Order {self.order_id} - Table {self.table.table_number} - {self.status.value} - ${self.get_total():.2f}" 

File: JAVA/lld-examples/medium/restaurant-management-system/src/main/java/com/systemdesign/lld/restaurant/Order.java
package com.systemdesign.lld.restaurant;

import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;
import java.util.UUID;

/**
 * Represents an order in the restaurant
 */
public class Order {
    private String orderId;
    private Table table;
    private Customer customer;
    private List<OrderItem> items;
    private OrderStatus status;
    private LocalDateTime orderTime;
    private LocalDateTime completionTime;
    private double totalAmount;
    
    public Order(Table table, Customer customer) {
        this.orderId = UUID.randomUUID().toString().substring(0, 8);
        this.table = table;
        this.customer = customer;
        this.items = new ArrayList<>();
        this.status = OrderStatus.PENDING;
        this.orderTime = LocalDateTime.now();
        this.totalAmount = 0.0;
    }
    
    public void addItem(MenuItem menuItem, int quantity) {
        OrderItem orderItem = new OrderItem(menuItem, quantity);
        items.add(orderItem);
        calculateTotal();
    }
    
    public void removeItem(MenuItem menuItem) {
        items.removeIf(item -> item.getMenuItem().equals(menuItem));
        calculateTotal();
    }
    
    public void updateItemQuantity(MenuItem menuItem, int newQuantity) {
        for (OrderItem item : items) {
            if (item.getMenuItem().equals(menuItem)) {
                item.setQuantity(newQuantity);
                break;
            }
        }
        calculateTotal();
    }
    
    private void calculateTotal() {
        totalAmount = items.stream()
                .mapToDouble(item -> item.getMenuItem().getPrice() * item.getQuantity())
                .sum();
    }
    
    public void confirm() {
        this.status = OrderStatus.CONFIRMED;
    }
    
    public void startPreparation() {
        this.status = OrderStatus.PREPARING;
    }
    
    public void markReady() {
        this.status = OrderStatus.READY;
    }
    
    public void complete() {
        this.status = OrderStatus.COMPLETED;
        this.completionTime = LocalDateTime.now();
    }
    
    public void cancel() {
        this.status = OrderStatus.CANCELLED;
    }
    
    public int getEstimatedPreparationTime() {
        return items.stream()
                .mapToInt(item -> item.getMenuItem().getPreparationTime())
                .max()
                .orElse(0);
    }
    
    // Getters
    public String getOrderId() { return orderId; }
    public Table getTable() { return table; }
    public Customer getCustomer() { return customer; }
    public List<OrderItem> getItems() { return items; }
    public OrderStatus getStatus() { return status; }
    public LocalDateTime getOrderTime() { return orderTime; }
    public LocalDateTime getCompletionTime() { return completionTime; }
    public double getTotalAmount() { return totalAmount; }
    
    @Override
    public String toString() {
        return String.format("Order %s - Table %d - Total: $%.2f - Status: %s",
                orderId, table.getTableNumber(), totalAmount, status);
    }
}

/**
 * Represents an item in an order
 */
class OrderItem {
    private MenuItem menuItem;
    private int quantity;
    
    public OrderItem(MenuItem menuItem, int quantity) {
        this.menuItem = menuItem;
        this.quantity = quantity;
    }
    
    public MenuItem getMenuItem() { return menuItem; }
    public int getQuantity() { return quantity; }
    public void setQuantity(int quantity) { this.quantity = quantity; }
    
    public double getSubtotal() {
        return menuItem.getPrice() * quantity;
    }
    
    @Override
    public String toString() {
        return String.format("%s x%d - $%.2f", 
                menuItem.getName(), quantity, getSubtotal());
    }
}

enum OrderStatus {
    PENDING,
    CONFIRMED,
    PREPARING,
    READY,
    COMPLETED,
    CANCELLED
} 